donkeycar/parts/pins.py

```diff
@@ -0,0 +1,1043 @@
+    """
+    PWM output pin using Rpi.GPIO/Jetson.GPIO
+    """
+    def __init__(self, pin_number: int, pin_scheme: str, frequency_hz: float = 50) -> None:
+        self.pin_number = pin_number
+        self.pin_scheme = gpio_pin_scheme[pin_scheme]
+        self.frequency = int(frequency_hz)
+        self.pwm = None
+        self._state = PinState.NOT_STARTED
+
+    def start(self, duty: float = 0) -> None:
+        if self.pwm is not None:
+            raise RuntimeError("Attempt to start PwmPinGpio that is already started.")
+        if duty < 0 or duty > 1:
+            raise ValueError("duty_cycle must be in range 0 to 1")
+        gpio_fn(self.pin_scheme, lambda: GPIO.setup(self.pin_number, GPIO.OUT))
+        self.pwm = gpio_fn(self.pin_scheme, lambda: GPIO.PWM(self.pin_number, self.frequency))
+        self.pwm.start(duty * 100)  # takes duty in range 0 to 100
+        self._state = duty
+
+    def stop(self) -> None:
+        if self.pwm is not None:
+            self.pwm.stop()
+            gpio_fn(self.pin_scheme, lambda: GPIO.cleanup(self.pin_number))
+        self._state = PinState.NOT_STARTED
+
+    def state(self) -> float:
+        return self._state
+
+    def duty_cycle(self, duty: float) -> None:
+        if duty < 0 or duty > 1:
+            raise ValueError("duty_cycle must be in range 0 to 1")
+        self.pwm.ChangeDutyCycle(duty * 100)  # takes duty of 0 to 100
+        self._state = duty
+
+
+#
+# ----- PCA9685 implementations -----
+#
+class PCA9685:
+    '''
+    Pin controller using PCA9685 boards.
+    This is used for most RC Cars.  This
+    driver can output ttl HIGH or LOW or
+    produce a duty cycle at the given frequency.
+    '''
+    def __init__(self, busnum: int, address: int, frequency: int):
+
+        import Adafruit_PCA9685
+        if busnum is not None:
+            from Adafruit_GPIO import I2C
+
+            # monkey-patch I2C driver to use our bus number
+            def get_bus():
+                return busnum
+
+            I2C.get_default_bus = get_bus
+        self.pwm = Adafruit_PCA9685.PCA9685(address=address)
+        self.pwm.set_pwm_freq(frequency)
+        self._frequency = frequency
+
+    def get_frequency(self):
+        return self._frequency
+
+    def set_high(self, channel: int):
+        self.pwm.set_pwm(channel, 4096, 0)
+
+    def set_low(self, channel: int):
+        self.pwm.set_pwm(channel, 0, 4096)
+
+    def set_duty_cycle(self, channel: int, duty_cycle: float):
+        if duty_cycle < 0 or duty_cycle > 1:
+            raise ValueError("duty_cycle must be in range 0 to 1")
+        if duty_cycle == 1:
+            self.set_high(channel)
+        elif duty_cycle == 0:
+            self.set_low(channel)
+        else:
+            # duty cycle is fraction of the 12 bits
+            pulse = int(4096 * duty_cycle)
+            try:
+                self.pwm.set_pwm(channel, 0, pulse)
+            except:
+                self.pwm.set_pwm(channel, 0, pulse)
```

Comment:
This looks spurious, if the `try` failed we should log an error, but not simply try again.

Reply:
I ported this.  Looks like a one-time retry.  I'll check with the original author why it is there, but I don't want to remove it without understanding that.  I can't really hurt leaving it, but it might hurt taking it out.

Reply:
I think it was Doug in d1efdfa4056cdbb0b6ae623397013dc9f625be74, and the code was in the PCA9685 before. I have taken this bit out of my fork a long time ago, as I saw regular issues where it was throwing from the `except` branch, because the I2C can sometimes produce a hick up. If a single pwm signal gets missed, we are just a missing a control for 1/vehicle_HZ seconds, so not too bad.

Reply:
@DocGarbanzo Is there a reason not to do this retry?

Reply:
Yes, if there is an I2C error, which is something I have observed on my car, then the non-exception handled retry might throw which will stop the vehicle loop. The `except` code should simply log an error, and when the vehicle loop comes around again, it will `try` to set it again.